apps/test/levelTests.js

```diff
@@ -11,32 +11,105 @@
+function loadSource(src) {
+  var deferred = new $.Deferred();
+  document.body.appendChild($('<script>', { src: src }).on('load', function () {
+    deferred.resolve();
+  })[0]);
+  return deferred;
+}
+describe('Level tests', function() {
+  var studioApp;
+  var originalRender;
+
+  before(function(done) {
+    this.timeout(15000);
+    // Load a bunch of droplet sources. We could potentially gate this on level.editCode,
+    // but that doesn't get us a lot since everything is run in a single session now.
+    loadSource('http://localhost:8001/jsinterpreter/acorn_interpreter.js')
+    .then(function () { return loadSource('http://localhost:8001/requirejs/full/require.js'); })
+    .then(function () { return loadSource('http://localhost:8001/ace/src-noconflict/ace.js'); })
+    .then(function () { return loadSource('http://localhost:8001/ace/src-noconflict/mode-javascript.js'); })
+    .then(function () { return loadSource('http://localhost:8001/ace/src-noconflict/ext-language_tools.js'); })
+    .then(function () { return loadSource('http://localhost:8001/droplet/droplet-full.js'); })
+    .then(function () {
+      assert(requirejs);
+      done();
+    });
+  });
+
+  beforeEach(function () {
+    testUtils.setupBlocklyFrame();
+    studioApp = testUtils.getStudioAppSingleton();
+
+    wrappedEventListener.attach();
+
+    // For some reason, svg rendering is taking a long time in phantomjs. None
+    // of these tests depend on that rendering actually happening.
+    originalRender = Blockly.BlockSvg.prototype.render;
+    Blockly.BlockSvg.prototype.render = function () {
+      this.block_.rendered = true;
+    };
+
+    if (window.Studio) {
+      var Projectile = require('@cdo/apps/studio/projectile');
+      Projectile.__resetIds();
+    }
+  });
+
+  getTestCollections().forEach(runTestCollection);
+
+  afterEach(function () {
+    var studioApp = require('@cdo/apps/StudioApp').singleton;
+    if (studioApp.editor && studioApp.editor.aceEditor &&
+        studioApp.editor.aceEditor.session &&
+        studioApp.editor.aceEditor.session.$mode &&
+        studioApp.editor.aceEditor.session.$mode.cleanup) {
+      studioApp.editor.aceEditor.session.$mode.cleanup();
+    }
+    wrappedEventListener.detach();
+    Blockly.BlockSvg.prototype.render = originalRender;
+
+    // Clean up some state that is meant to be per level. This is an issue
+    // because we're using the same window for all tests.
+    if (window.Maze) {
+      window.Maze.bee = null;
+      window.Maze.wordSearch = null;
+    }
+    if (window.Studio) {
+      window.Studio.customLogic = null;
+      window.Studio.interpreter = null;
+    }
+  });
 });
 
 // Get all json files under directory path
-function getTestCollections (directory) {
-  var files = wrench.readdirSyncRecursive(directory);
+function getTestCollections () {
+  // require-globify transform
+  var files = require('./solutions/**/*.js', {hash: 'path'});
```

Comment:
I agree. I think doing something like that becomes a lot easier with our
new setup vs. how it worked before.

On Mon, May 11, 2015 at 7:19 PM, Will Jordan <notifications@github.com>
wrote:

> In apps/test/levelTests.js
> <https://github.com/code-dot-org/code-dot-org/pull/2376#discussion_r30100198>
> :
>
> >  });
> >
> >  // Get all json files under directory path
> > -function getTestCollections (directory) {
> > -  var files = wrench.readdirSyncRecursive(directory);
> > +function getTestCollections () {
> > +  // require-globify transform
> > +  var files = require('./solutions/**/*.js', {hash: 'path'});
>
> Note for future- I've been thinking that if it's not too much work, at
> some point it would be nice to invert this include pattern, so that instead
> of doing the require-glob within this test code to get all tests, each
> /solutions/.../test.js file would serve as a self-contained test that
> requires all of its necessary test dependencies.
>
> That way, each individual .js test file could be tested independently of
> any other, and the globbing-logic would be delegated to the test runner
> (e.g. mocha test/**/*.js).
>
> Not necessary work, just something to keep in mind if we wanted to make
> this test codebase even more unix-y modular (thinking about the
> browserify-author's how I write tests for node and the browser
> <http://substack.net/how_I_write_tests_for_node_and_the_browser> here).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/2376/files#r30100198>.
>